File: ios/Runner/MotionBridge.swift
import CoreMotion
import Flutter
import UIKit

/// Streams device attitude to Dart.
///
/// Core Motion already fuses the gyroscope, accelerometer and magnetometer and
/// hands back an attitude referenced to true north, which is far better than
/// anything assembled from raw sensor streams on the Dart side. The whole
/// point of this file is to get at `xTrueNorthZVertical` and hand it over.
class MotionBridge: NSObject, FlutterStreamHandler {
    static let channelName = "cosmolabe/orientation"

    private let motionManager = CMMotionManager()
    private var eventSink: FlutterEventSink?

    /// Core Motion's true-north frame has X towards north and Z up, which makes
    /// Y point west. The rest of the app works in East-North-Up, so every
    /// sample is rotated a quarter turn about the vertical before it leaves.
    private static let nwuToEnu = (w: (2.0).squareRoot() / 2, x: 0.0, y: 0.0, z: (2.0).squareRoot() / 2)

    static func register(with registrar: FlutterPluginRegistrar) {
        let bridge = MotionBridge()
        let channel = FlutterEventChannel(
            name: channelName,
            binaryMessenger: registrar.messenger()
        )
        channel.setStreamHandler(bridge)

        let methods = FlutterMethodChannel(
            name: "\(channelName)/control",
            binaryMessenger: registrar.messenger()
        )
        methods.setMethodCallHandler { call, result in
            switch call.method {
            case "isAvailable":
                result(bridge.motionManager.isDeviceMotionAvailable)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }

    func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        guard motionManager.isDeviceMotionAvailable else {
            return FlutterError(
                code: "unavailable",
                message: "This device has no motion sensors",
                details: nil
            )
        }

        eventSink = events

        // 60 Hz matches the display; asking for more only burns battery.
        motionManager.deviceMotionUpdateInterval = 1.0 / 60.0
        motionManager.startDeviceMotionUpdates(
            using: .xTrueNorthZVertical,
            to: .main
        ) { [weak self] motion, error in
            guard let self else { return }

            if let error {
                events(FlutterError(
                    code: "motion_failed",
                    message: error.localizedDescription,
                    details: nil
                ))
                return
            }
            guard let motion else { return }

            let q = motion.attitude.quaternion
            let rotated = Self.applyFrameCorrection(w: q.w, x: q.x, y: q.y, z: q.z)

            events([
                "w": rotated.w,
                "x": rotated.x,
                "y": rotated.y,
                "z": rotated.z,
                // Core Motion reports calibration state rather than an angle;
                // map it onto the same "how many degrees might this be wrong"
                // scale that Android gives us.
                "accuracy": Self.headingAccuracy(motion.magneticField.accuracy),
                "trueNorth": true,
            ])
        }

        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        motionManager.stopDeviceMotionUpdates()
        eventSink = nil
        return nil
    }

    /// Left-multiplies the sample by the north-west-up to east-north-up
    /// rotation.
    private static func applyFrameCorrection(
        w: Double, x: Double, y: Double, z: Double
    ) -> (w: Double, x: Double, y: Double, z: Double) {
        let c = nwuToEnu
        return (
            w: c.w * w - c.x * x - c.y * y - c.z * z,
            x: c.w * x + c.x * w + c.y * z - c.z * y,
            y: c.w * y - c.x * z + c.y * w + c.z * x,
            z: c.w * z + c.x * y - c.y * x + c.z * w
        )
    }

    private static func headingAccuracy(
        _ accuracy: CMMagneticFieldCalibrationAccuracy
    ) -> Double {
        switch accuracy {
        case .high: return 5.0
        case .medium: return 15.0
        case .low: return 35.0
        default: return 180.0
        }
    }
}

File: android/app/src/main/kotlin/ai/azenha/cosmolabe/MotionBridge.kt
package ai.azenha.cosmolabe

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel

/**
 * Streams device attitude to Dart.
 *
 * TYPE_ROTATION_VECTOR is the platform's own sensor fusion, already blending
 * gyroscope, accelerometer and magnetometer. Its world frame is exactly the
 * East-North-Up the app works in, so unlike iOS no frame correction is needed
 * here — but it is referenced to *magnetic* north, so the Dart side has to
 * apply the declination itself.
 */
class MotionBridge(context: Context) :
    EventChannel.StreamHandler, SensorEventListener {

    companion object {
        const val CHANNEL_NAME = "cosmolabe/orientation"

        fun register(messenger: BinaryMessenger, context: Context) {
            val bridge = MotionBridge(context)
            EventChannel(messenger, CHANNEL_NAME).setStreamHandler(bridge)
            MethodChannel(messenger, "$CHANNEL_NAME/control")
                .setMethodCallHandler { call, result ->
                    when (call.method) {
                        "isAvailable" -> result.success(bridge.isAvailable())
                        else -> result.notImplemented()
                    }
                }
        }
    }

    private val sensorManager =
        context.getSystemService(Context.SENSOR_SERVICE) as SensorManager

    private var eventSink: EventChannel.EventSink? = null

    /** Reused across every sample; allocating in a 60 Hz callback would hand
     * the garbage collector a steady stream of short-lived arrays. */
    private val quaternion = FloatArray(4)

    private var headingAccuracy = 180.0

    private fun rotationSensor(): Sensor? =
        sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)
            ?: sensorManager.getDefaultSensor(Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR)

    private fun isAvailable(): Boolean = rotationSensor() != null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events

        val sensor = rotationSensor()
        if (sensor == null) {
            events?.error("unavailable", "This device has no rotation sensor", null)
            return
        }

        sensorManager.registerListener(this, sensor, SensorManager.SENSOR_DELAY_GAME)
        // The magnetometer's own accuracy arrives separately, through
        // onAccuracyChanged; watch it so the calibration prompt can appear.
        sensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)?.let {
            sensorManager.registerListener(this, it, SensorManager.SENSOR_DELAY_NORMAL)
        }
    }

    override fun onCancel(arguments: Any?) {
        sensorManager.unregisterListener(this)
        eventSink = null
    }

    override fun onSensorChanged(event: SensorEvent?) {
        val sink = eventSink ?: return
        if (event == null) return
        if (event.sensor.type != Sensor.TYPE_ROTATION_VECTOR &&
            event.sensor.type != Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR
        ) {
            return
        }

        // Returns [w, x, y, z] in the East-North-Up world frame.
        SensorManager.getQuaternionFromVector(quaternion, event.values)

        sink.success(
            mapOf(
                "w" to quaternion[0].toDouble(),
                "x" to quaternion[1].toDouble(),
                "y" to quaternion[2].toDouble(),
                "z" to quaternion[3].toDouble(),
                "accuracy" to headingAccuracy,
                // Magnetic north, not true north.
                "trueNorth" to false,
            )
        )
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
        if (sensor?.type != Sensor.TYPE_MAGNETIC_FIELD) return
        headingAccuracy = when (accuracy) {
            SensorManager.SENSOR_STATUS_ACCURACY_HIGH -> 5.0
            SensorManager.SENSOR_STATUS_ACCURACY_MEDIUM -> 15.0
            SensorManager.SENSOR_STATUS_ACCURACY_LOW -> 35.0
            else -> 180.0
        }
    }
}
